include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -3139,16 +3137,14 @@ namespace internal
       const unsigned int                     subface_index,
       std::integral_constant<bool, false>)
     {
-      // TODO. Need to make sure hanging nodes work, i.e call
-      // create_eval_tensor_product with correct direction.
```

Comment:
Is it really enough to treat the hanging nodes only by `FEFaceEvaluation`? What does happen you use `FEEvaluation`? The contributions of constrained DoFs are not added to the constraining ones.

Reply:
The shape data will be the same for `FEEvaluation` since each cell is still evaluated in the entire reference cell. For `FEFaceEvaluation` we instead have to evaluate either between coordinates `[0, 0.5]x[0, 0.5]`, or `[0, 0.5]x[0.5, 1]`, etc, in 3D. It's `[0, 0.5]` or `[0.5, 1]` in 2D.

As far as I know, the constrained dofs are just ignored when calling `FEEvaluation.distribute_local_to_global(..)`, and only updated if you call `constraints.distribute(..)`, but you probably know this a lot better than me.

Reply:
> As far as I know, the constrained dofs are just ignored when calling FEEvaluation.distribute_local_to_global(..), and only updated if you call constraints.distribute(..), but you probably know this a lot better than me.

They are not ignored, but their contributions are added to the constraining DoFs. Your test should give you different results depending on `DoFTools::make_hanging_node_constraints` was called or not? Could you check?

Reply:
> Your test should give you different results depending on `DoFTools::make_hanging_node_constraints` was called or not?

It does not make a difference if I call it or not. Maybe I need a different test? I also does not manage to finish on the CI check since the matrix assembly is so damn slow for some reason.

Reply:
> It does not make a difference if I call it or not. Maybe I need a different test?

There should be a difference. At the moment, you are only comparing matrix-free and matrix-based results. Those should be consistent if correctly implemented. If you add hanging-node constraints, the l2 norm of the vectors should differ.

Reply:
Ah, yes ofc. It does make a difference!